std/experimental/ndslice/selection.d

```diff
@@ -747,36 +747,172 @@ pure nothrow unittest
+    auto slice = iotaSlice(3, 4)
+        .allReversed
+        .reshape(-1, 3);
+    assert(slice ==
+        [[11, 10, 9],
+         [ 8,  7, 6],
+         [ 5,  4, 3],
+         [ 2,  1, 0]]);
+}
+
+pure unittest
+{
+    import std.experimental.ndslice.slice;
+    import std.experimental.ndslice.iteration : reversed;
+    import std.array : array;
+
+    auto reshape2(S, size_t M)(S slice, size_t[M] lengths...)
+    {
+        // Tries to reshape without allocation
+        try return slice.reshape(lengths);
+        catch (ReshapeException e)
+            // Allocates
+            return slice.slice.reshape(lengths);
+    }
+
+    auto slice =
+        [0, 1,  2,  3,
+         4, 5,  6,  7,
+         8, 9, 10, 11]
+        .sliced(3, 4)
+        .reversed!0;
+
+    assert(reshape2(slice, 4, 3) ==
+        [[ 8, 9, 10],
+         [11, 4,  5],
+         [ 6, 7,  0],
+         [ 1, 2,  3]]);
+}
+
+@safe pure unittest
+{
+    import std.experimental.ndslice.iteration : allReversed;
+    auto slice = iotaSlice(1, 1, 3, 2, 1, 2, 1).allReversed;
+    assert(slice.reshape(1, -1, 1, 1, 3, 1) ==
+        [[[[[[11], [10], [9]]]],
+          [[[[ 8], [ 7], [6]]]],
+          [[[[ 5], [ 4], [3]]]],
+          [[[[ 2], [ 1], [0]]]]]]);
+}
+
+// Issue 15919
+unittest
+{
+    assert(iotaSlice(3, 4, 5, 6, 7).pack!2.reshape(4, 3, 5)[0, 0, 0].shape == cast(size_t[2])[6, 7]);
+}
+
+@safe pure unittest
+{
+    import std.experimental.ndslice.slice;
+    import std.range : iota;
+    import std.exception : assertThrown;
+
+    auto e = 1.iotaSlice(1);
+    // resize to the wrong dimension
+    assertThrown!ReshapeException(e.reshape(2));
+    e.popFront;
+    // test with an empty slice
+    assertThrown!ReshapeException(e.reshape(1));
+}
+
+unittest
+{
+    auto pElements = iotaSlice(3, 4, 5, 6, 7)
+        .pack!2
+        .byElement();
+    assert(pElements[0][0] == iotaSlice(7));
+    assert(pElements[$-1][$-1] == iotaSlice([7], 2513));
+}
+
+/++
+Error codes for $(LREF reshape).
++/
+enum ReshapeError
+{
+    /// No error
+    none,
+    /// Slice should be not empty
+    empty,
+    /// Total element count should be the same
+    total,
+    /// Structure is incompatible with new shape
+    incompatible,
+}
+
 /++
 Returns a new slice for the same data with different dimensions.
 
 Params:
     slice = slice to be reshaped
     lengths = list of new dimensions. One of the lengths can be set to `-1`.
         In this case, the corresponding dimension is inferable.
+    err = $(LREF ReshapeError) code
 Returns:
     reshaped slice
-Throws:
-    $(LREF ReshapeException) if the slice cannot be reshaped with the input lengths.
 +/
 Slice!(M, Range)
     reshape
         (size_t N, Range, size_t M)
-        (Slice!(N, Range) slice, size_t[M] lengths...)
+        (Slice!(N, Range) slice, size_t[M] lengths, ref int err)
 {
     mixin _DefineRet;
     foreach (i; Iota!(0, ret.N))
         ret._lengths[i] = lengths[i];
 
+    goto B;
+R:
+        return ret;
```

Comment:
I think you're doing this (as opposed to just issuing `return ret;` directly) to save code size? Please add an explanatory comment, otherwise this will jump at anyone doing refactoring.